### PR TITLE
Make DocHande !Sync

### PR DESCRIPTION
**Pull request by teohhanhui**
Enforce at the type level that DocHandle is not to be shared across threads.

/cc @gterzian: https://github.com/bowtieworks/automerge_orm/issues/3#issuecomment-1803534732

### CODE REVIEW

File: src/dochandle.rs [hunk 1]
```diff
@@ -30,6 +32,7 @@ pub struct DocHandle {
     /// that doesn't require a mutabale reference to the handle.
     /// Note: the mutex is not shared between clones of the same handle.
     last_heads: Mutex<Vec<ChangeHash>>,
+    _not_sync: PhantomData<Cell<&'static ()>>,
```
Comment (teohhanhui) [inline src/dochandle.rs:35]: ref: https://doc.rust-lang.org/nomicon/phantom-data.html#table-of-phantomdata-patterns

File: examples/distributed_bakery.rs [hunk 2]
```diff
@@ -60,8 +61,8 @@ async fn increment_output(doc_handle: &DocHandle) -> Result<u32, ()> {
     Ok(latest)
 }
 
-async fn run_bakery_algorithm(doc_handle: &DocHandle, customer_id: &String) {
-    let (our_number, closing) = doc_handle.with_doc_mut(|doc| {
+async fn run_bakery_algorithm(doc_handle: &Mutex<DocHandle>, customer_id: &String) {
+    let (our_number, closing) = doc_handle.lock().await.with_doc_mut(|doc| {
```
Comment (teohhanhui) [inline examples/distributed_bakery.rs:65]: I think the locking behaviour here and elsewhere is probably wrong, but I'm not sure what to do...

@gterzian if you could advise?

**Comment by gterzian** [comment]
@teohhanhui Thanks for the PR!

couple of questions after a quick look:

1. Can we not just do `impl !Sync for DocHandle`?
2. Why the mutex around the handle in the tests?

**Comment by teohhanhui** [comment]
> Can we not just do `impl !Sync for DocHandle`?

`negative_impls` is still unstable: https://github.com/rust-lang/rust/issues/68318

The mutex is necessary because of https://github.com/tokio-rs/axum/issues/1463

**Comment by gterzian** [comment]
> The mutex is necessary because of https://github.com/tokio-rs/axum/issues/1463

Ok I see, [tokio](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.spawn) only requires futures to be `Send`. 

I think for now I would rather keep the doc handle as it is, and have a kind of loose convention that if relying on `changed` and sharing it with an arc could be confusing. 

A good example would be https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html, which also is `Sync`, even though sharing it across task is probably not a good idea.

**Comment by teohhanhui** [comment]
> A good example would be https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html, which also is `Sync`, even though sharing it across task is probably not a good idea.

Actually it's documented to be fine: https://docs.rs/tokio/latest/tokio/sync/watch/index.html#thread-safety

**Comment by gterzian** [comment]
> using it concurrently is technically correct but what you probably want to do is to clone it.

Exactly

**Pull request closed without merging**